client/state/data-layer/convert-snake-case-to-camel-case.ts

```diff
@@ -0,0 +1,37 @@
+type SnakeToCamel< S extends string > = S extends `${ infer T }_${ infer U }`
+	? `${ T }${ Capitalize< SnakeToCamel< U > > }`
+	: S;
+
+type ConvertToCamelCase< T > = T extends object
+	? T extends Array< infer U >
+		? Array< ConvertToCamelCase< U > >
+		: {
+				[ K in keyof T as SnakeToCamel< K & string > ]: ConvertToCamelCase< T[ K ] >;
+		  }
+	: T;
+
+export function convertSnakeCaseToCamelCase< T >( obj: T ): ConvertToCamelCase< T > {
+	if ( typeof obj !== 'object' || obj === null ) {
+		return obj as ConvertToCamelCase< T >;
+	}
+
+	if ( Array.isArray( obj ) ) {
+		return obj.map( ( item ) =>
+			convertSnakeCaseToCamelCase( item )
+		) as unknown as ConvertToCamelCase< T >;
+	}
+
+	const toCamelCase = ( str: string ) =>
+		str.replace( /_([a-z])/g, ( _, letter ) => letter.toUpperCase() );
+
+	const newObj: any = {};
+
+	for ( const key in obj ) {
+		if ( obj.hasOwnProperty( key ) ) {
+			const camelCaseKey = toCamelCase( key );
+			newObj[ camelCaseKey ] = convertSnakeCaseToCamelCase( obj[ key ] );
+		}
+	}
+
+	return newObj as ConvertToCamelCase< T >;
+}
```

Comment:
This code is a bit convoluted, and maybe that's required to handle edge cases. Still I'm wondering if we need those two types, and if we could go for something simpler that is also easier to understand:

```javascript
function camelCase( string: string ): string {
	return string.replace( /_([a-z])/g, ( _, letter ) => letter.toUpperCase() );
}

export function convertSnakeCaseToCamelCase<T extends Record<string, any>>( object: T ): T {
	const newObject: Record<string, any> = {};

	Object.entries( object ).forEach( ( [ key, value ] ) => {
		if ( typeof value === 'object' ) {
			value = convertSnakeCaseToCamelCase( value );
		} else if ( Array.isArray( value ) ) {
			value = value.map( item =>
				typeof item === 'object' ? convertSnakeCaseToCamelCase( item ) : item
			);
		}

		newObject[camelCase( key )] = value;
	} );

	return newObject as T;
}
```
This code passes the unit tests, and `yarn tsc -p client/tsconfig.json --noEmit` doesn't trigger any warning.

Reply:
It's much easier to show the difference with the help of video, so I recorded it :)
https://github.com/Automattic/wp-calypso/assets/5598437/afa2b22b-2706-4304-a146-efd8f8e809c9